presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -542,6 +542,19 @@ public PlanOptimizers(
                 predicatePushDown,
                 simplifyRowExpressionOptimizer); // Should always run simplifyOptimizer after predicatePushDown
 
+        builder.add(new OptimizeMixedDistinctAggregations(metadata));
+        builder.add(new IterativeOptimizer(
+                ruleStats,
+                statsCalculator,
+                estimatedExchangesCostCalculator,
+                ImmutableSet.of(
+                        new CreatePartialTopN(),
+                        new PushTopNThroughUnion())));
+
+        // We do a single pass, and assign `statsEquivalentPlanNode` to each node.
+        // After this step, nodes with same `statsEquivalentPlanNode` will share same history based statistics.
+        builder.add(new StatsRecordingPlanOptimizer(optimizerStats, new HistoricalStatisticsEquivalentPlanMarkingOptimizer()));
```

Comment:
Reordering the optimizer rule orders are in general very dangerous. Have we done enough shadow testing to make sure it will cause regression?

Reply:
I did so because these rules handle very different nodes(aggregations/unions vs joins). The benefit to move the rules is small(we get history for a bit more nodes in cases of union+topn. 

If you still think its dangerous, I will move it back

Reply:
this looks pretty safe for reordering.  to confirm you can run a verifier test and see if any plans change with the different orders. (you'll have to compare the plans after the run using the logging tables).

Reply:
@rschlussel I ran the verifier, only 3 queries had different plans, all of them were due to sort + window plan optimizer change you made.